File: fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.java
/*
 * Copyright 2016 Shazam Entertainment Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.shazam.fork.runner.listeners;

import com.android.ddmlib.testrunner.TestIdentifier;
import com.shazam.fork.batch.tasks.TestTask;
import com.shazam.fork.model.Device;
import com.shazam.fork.model.Pool;
import com.shazam.fork.model.TestCaseEvent;
import com.shazam.fork.model.TestCaseEventFactory;
import com.shazam.fork.runner.ProgressReporter;
import com.shazam.fork.system.io.FileManager;
import com.shazam.fork.system.io.FileType;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.Map;
import java.util.Queue;

import javax.annotation.Nonnull;

import static com.google.common.base.Preconditions.checkNotNull;

public class RetryListener extends NoOpITestRunListener {

    private static final Logger logger = LoggerFactory.getLogger(RetryListener.class);

    @Nonnull
    private final Device device;
    @Nonnull
    private final Queue<TestTask> queueOfTestsInPool;
    @Nonnull
    private final TestTask currentTestCaseEvent;
    private ProgressReporter progressReporter;
    private FileManager fileManager;
    private Pool pool;
    private final TestCaseEventFactory testCaseEventFactory;

    public RetryListener(@Nonnull Pool pool, @Nonnull Device device,
                         @Nonnull Queue<TestTask> queueOfTestsInPool,
                         @Nonnull TestTask currentTestCaseEvent,
                         @Nonnull ProgressReporter progressReporter,
                         FileManager fileManager,
                         TestCaseEventFactory testCaseEventFactory) {
        checkNotNull(device);
        checkNotNull(queueOfTestsInPool);
        checkNotNull(currentTestCaseEvent);
        checkNotNull(progressReporter);
        checkNotNull(pool);
        this.device = device;
        this.queueOfTestsInPool = queueOfTestsInPool;
        this.currentTestCaseEvent = currentTestCaseEvent;
        this.progressReporter = progressReporter;
        this.pool = pool;
        this.fileManager = fileManager;
        this.testCaseEventFactory = testCaseEventFactory;
    }

    @Override
    public void testFailed(TestIdentifier test, String trace) {
        progressReporter.recordFailedTestCase(pool, testCaseEventFactory.newTestCase(test));

        if (progressReporter.requestRetry(pool, testCaseEventFactory.newTestCase(test))) {
            queueOfTestsInPool.add(currentTestCaseEvent);
            logger.info("Test " + test.toString() + " enqueued again into pool:" + pool.getName());
            removeFailureTraceFiles(test);
        } else {
            logger.info("Test " + test.toString() + " failed on device " + device.getSafeSerial() + " but retry is not allowed.");
        }
    }


    private void removeFailureTraceFiles(TestIdentifier test) {
        final File file = fileManager.getFile(FileType.TEST, pool.getName(), device.getSafeSerial(), test);
        boolean deleted = file.delete();
        if (!deleted) {
            logger.warn("Failed to remove file  " + file.getAbsoluteFile() + " for a failed but enqueued again test");
        }
    }
}

File: fork-runner/src/main/java/com/shazam/fork/runner/listeners/RetryListener.kt
package com.shazam.fork.runner.listeners

import com.android.ddmlib.testrunner.TestIdentifier
import com.android.ddmlib.testrunner.TestResult
import com.android.ddmlib.testrunner.TestRunResult
import com.shazam.fork.batch.tasks.TestTask
import com.shazam.fork.model.Device
import com.shazam.fork.model.Pool
import com.shazam.fork.model.TestCaseEvent
import com.shazam.fork.model.TestCaseEventFactory
import com.shazam.fork.runner.ProgressReporter
import com.shazam.fork.system.io.FileManager
import com.shazam.fork.system.io.FileType
import org.ietf.jgss.GSSException.FAILURE
import org.slf4j.LoggerFactory
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.log

class RetryListener(private val pool: Pool,
                    private val device: Device,
                    private val queueOfTestsInPool: Queue<TestTask>,
                    private val progressReporter: ProgressReporter,
                    private val testTask: TestTask,
                    private val fileManager: FileManager,
                    private val testCaseEventFactory: TestCaseEventFactory) : NoOpITestRunListener() {

    private val logger = LoggerFactory.getLogger(RetryListener::class.java)

    private val testResults = TestRunResult()

    override fun testStarted(test: TestIdentifier) {
        testResults.testStarted(test)
    }

    override fun testFailed(test: TestIdentifier, trace: String) {
        testResults.testFailed(test, trace)
    }

    override fun testAssumptionFailure(test: TestIdentifier, trace: String) {
        testResults.testAssumptionFailure(test, trace)
    }

    override fun testRunStarted(runName: String, testCount: Int) {
        testResults.testRunStarted(runName, testCount)
    }

    override fun testRunStopped(elapsedTime: Long) {
        testResults.testRunStopped(elapsedTime)
    }

    override fun testIgnored(test: TestIdentifier) {
        testResults.testIgnored(test)
    }

    override fun testEnded(test: TestIdentifier, testMetrics: MutableMap<String, String>) {
        testResults.testEnded(test, testMetrics)
    }

    override fun testRunFailed(errorMessage: String) {
        testResults.testRunFailed(errorMessage)
    }

    override fun testRunEnded(elapsedTime: Long, runMetrics: Map<String, String>) {
        restartFailedTests()
    }

    private fun restartFailedTests() {
        val allFinished = testResults.testResults.all { isSuccessful(it.value) }
        when (testTask) {
            is TestTask.SingleTestTask -> {
                if (!allFinished) {
                    restartFailedTest(testTask.event)
                }
            }
            is TestTask.MultiTestTask -> {
                if (!allFinished || testResults.testResults.size != testTask.list.size) {
                    restartMultiTestTask(testTask)
                }
            }
        }
    }

    private fun TestCaseEvent.identifier(): TestIdentifier {
        return TestIdentifier(testClass, testMethod)
    }

    private fun restartMultiTestTask(testTask: TestTask.MultiTestTask) {
        val tests = testTask.list
        val results = testResults.testResults

        val notExecuted = tests.filterNot { results.containsKey(it.identifier()) }

        val failed = results.filterValues { !isSuccessful(it) }

        failed.keys.map {
            testCaseEventFactory.newTestCase(it)
        }.forEach { restartFailedTest(it) }

        notExecuted.map {
            TestTask.SingleTestTask(it)
        }.forEach {
            queueOfTestsInPool.add(it)
        }
    }

    private fun isSuccessful(it: TestResult) =
            it.status == TestResult.TestStatus.PASSED || it.status == TestResult.TestStatus.IGNORED

    private fun restartFailedTest(event: TestCaseEvent) {
        val test = TestIdentifier(event.testClass, event.testMethod)
        progressReporter.recordFailedTestCase(pool, event)

        if (progressReporter.requestRetry(pool, event)) {
            queueOfTestsInPool.add(TestTask.SingleTestTask(event))
            logger.info("Test " + test.toString() + " enqueued again into pool:" + pool.name)
            removeFailureTraceFiles(test)
        } else {
            logger.info("Test " + test.toString() + " failed on device " + device.safeSerial + " but retry is not allowed.")
        }
    }


    private fun removeFailureTraceFiles(test: TestIdentifier) {
        val file = fileManager.getFile(FileType.TEST, pool.name, device.safeSerial, test)
        val deleted = file.delete()
        if (!deleted) {
            logger.warn("Failed to remove file  " + file.absoluteFile + " for a failed but enqueued again test")
        }
    }
}
